File: core/nf/n3iwf/esp/esp.go
// Copyright (c) 2026 MakeMyTechnology. All rights reserved.
//
// Package esp implements RFC 4303 Encapsulating Security Payload
// for the N3IWF NWu data plane (TS 24.502 §7.4 user-plane SAs and
// the signalling SA over which NAS rides).
//
// Cipher suite scope (matching what the IKEv2 handler negotiates in
// CREATE_CHILD_SA per nf/n3iwf/ikev2): AES-CBC-256 for confidentiality
// + HMAC-SHA-256-128 (RFC 4868 §2.1) for integrity, in tunnel mode
// (Next Header = 4 / IPv4 or 41 / IPv6 per IANA "Assigned Internet
// Protocol Numbers"). AEAD modes are out of scope for this phase.
//
// Wire layout (RFC 4303 §2 Figure 1, verbatim):
//
//	SPI (4)
//	Sequence Number (4)
//	Payload Data — IV (block_size, AES-CBC) || Ciphertext
//	Padding (0..255) | Pad Length (1) | Next Header (1)   ⟵ encrypted
//	ICV (variable)                                        ⟵ over SPI..ciphertext
//
// Authoritative spec: RFC 4303 (specs/ietf/rfc4303.txt).
package esp

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/binary"
	"errors"
	"fmt"
)

// Header lengths.
const (
	HeaderLen = 8 // SPI(4) + SeqNum(4)
	IVLen     = 16
	ICVLen    = 16 // HMAC-SHA-256-128 truncated to 128 bits (RFC 4868 §2.1)

	// NextHdr IANA "Assigned Internet Protocol Numbers" values used in
	// tunnel mode — the encapsulated inner packet's IP version.
	NextHdrIPv4 uint8 = 4
	NextHdrIPv6 uint8 = 41
)

// SA holds the per-direction key + state needed to encap or decap
// ESP packets. One SA per direction — outbound ESP uses an SA with
// the responder→initiator (or initiator→responder, mirror) keys
// depending on which side this is.
//
// SeqOut starts at 1 — RFC 4303 §3.3.3: "The sender's counter is
// initialized to 0 when an SA is established. The sender increments
// the sequence number ... counter for this SA and inserts the
// low-order 32 bits of the value into the Sequence Number field."
// First transmitted Sequence Number is 1.
type SA struct {
	SPI       uint32
	EncrKey   []byte // AES-256 key (32 octets)
	IntegKey  []byte // HMAC-SHA-256 key (32 octets per RFC 4868 §2.1)
	SeqOut    uint32 // last sent seq num
	SeqInWin  *replayWindow
}

// NewSA builds an outbound or inbound SA with a 64-packet replay
// window per RFC 4303 §3.4.3 ("a minimum window size of 32 packets
// MUST be supported, but a window size of 64 is preferred and
// SHOULD be employed as the default").
func NewSA(spi uint32, encrKey, integKey []byte) (*SA, error) {
	if len(encrKey) != 32 {
		return nil, fmt.Errorf("esp: encr key length %d != 32 (AES-256)", len(encrKey))
	}
	if len(integKey) != sha256.Size {
		return nil, fmt.Errorf("esp: integ key length %d != %d (HMAC-SHA-256, RFC 4868 §2.1)",
			len(integKey), sha256.Size)
	}
	return &SA{
		SPI:      spi,
		EncrKey:  append([]byte(nil), encrKey...),
		IntegKey: append([]byte(nil), integKey...),
		SeqInWin: newReplayWindow(64),
	}, nil
}

// Encap wraps inner (a complete IPv4 or IPv6 packet) in an ESP
// packet using this SA's outbound state. nextHdr selects the inner
// payload's IP version (NextHdrIPv4 / NextHdrIPv6).
//
// Returns the ESP packet ready to drop into a UDP datagram (NAT-T,
// RFC 3948) or a raw IP socket (protocol 50, RFC 4303 §2).
func (s *SA) Encap(inner []byte, nextHdr uint8) ([]byte, error) {
	// Bump seq num + check overflow per §3.3.3.
	if s.SeqOut == 0xFFFFFFFF {
		return nil, errors.New("esp: outbound sequence number wrapped (RFC 4303 §3.3.3 — must rekey)")
	}
	// Generate a fresh, unpredictable IV — required for AES-CBC's
	// IND-CPA security. crypto/rand satisfies the cryptographic-PRNG
	// requirement.
	iv := make([]byte, IVLen)
	if _, err := rand.Read(iv); err != nil {
		return nil, err
	}
	s.SeqOut++
	return buildPacket(s.SPI, s.SeqOut, s.EncrKey, s.IntegKey, iv, inner, nextHdr)
}

// encapWithIV is the deterministic-IV variant used only by tests so
// the Go and Python sides can lock to a byte-exact hex vector.
// Production code MUST call Encap, which generates a fresh random IV
// per RFC 4303 §3.3 ("the IV [...] MUST be chosen ... to ensure that
// no two packets are encrypted with the same IV").
func (s *SA) encapWithIV(inner []byte, nextHdr uint8, iv []byte, seq uint32) ([]byte, error) {
	if len(iv) != IVLen {
		return nil, fmt.Errorf("esp: test IV length %d != %d", len(iv), IVLen)
	}
	return buildPacket(s.SPI, seq, s.EncrKey, s.IntegKey, iv, inner, nextHdr)
}

// buildPacket is the pure transformation: (inputs) → wire bytes. No
// state mutation, no IV generation, no seq-num bump — caller owns
// both. Shared by Encap (production path) and encapWithIV (test path).
func buildPacket(
	spi, seq uint32,
	encrKey, integKey, iv, inner []byte,
	nextHdr uint8,
) ([]byte, error) {
	if len(inner) == 0 {
		return nil, errors.New("esp: inner payload empty")
	}
	// Plaintext = inner | padding | padlen | nextHdr.
	// Total length must be multiple of AES block size (16). Pad
	// pattern per RFC 4303 §2.4: 1, 2, 3, ..., padlen.
	tailLen := 2 // PadLen + NextHdr
	rem := (len(inner) + tailLen) % aes.BlockSize
	padLen := 0
	if rem != 0 {
		padLen = aes.BlockSize - rem
	}
	if padLen > 255 {
		return nil, fmt.Errorf("esp: pad length %d > 255 (RFC 4303 §2.4)", padLen)
	}
	plaintext := make([]byte, len(inner)+padLen+tailLen)
	copy(plaintext, inner)
	for i := 0; i < padLen; i++ {
		plaintext[len(inner)+i] = byte(i + 1) // §2.4 pattern
	}
	plaintext[len(plaintext)-2] = byte(padLen)
	plaintext[len(plaintext)-1] = nextHdr

	block, err := aes.NewCipher(encrKey)
	if err != nil {
		return nil, err
	}
	ciphertext := make([]byte, len(plaintext))
	cipher.NewCBCEncrypter(block, iv).CryptBlocks(ciphertext, plaintext)

	// Assemble: SPI | Seq | IV | Ciphertext (we'll append ICV).
	pkt := make([]byte, 0, HeaderLen+IVLen+len(ciphertext)+ICVLen)
	pkt = appendU32(pkt, spi)
	pkt = appendU32(pkt, seq)
	pkt = append(pkt, iv...)
	pkt = append(pkt, ciphertext...)

	// ICV = HMAC-SHA-256(IntegKey, SPI || Seq || IV || Ciphertext)
	// truncated to 128 bits per RFC 4303 §2.8 + RFC 4868 §2.1.
	mac := hmac.New(sha256.New, integKey)
	mac.Write(pkt) // covers SPI..Ciphertext exactly per §2.8
	icv := mac.Sum(nil)[:ICVLen]
	pkt = append(pkt, icv...)
	return pkt, nil
}

// Decap unwraps an ESP packet, verifies the ICV (constant-time),
// runs anti-replay (§3.4.3), AES-CBC-decrypts, strips padding, and
// returns the inner IP packet plus its NextHdr (NextHdrIPv4 / IPv6).
//
// SPI must match s.SPI — caller is expected to have demuxed by SPI
// before calling Decap.
func (s *SA) Decap(pkt []byte) ([]byte, uint8, error) {
	if len(pkt) < HeaderLen+IVLen+aes.BlockSize+ICVLen {
		return nil, 0, fmt.Errorf("esp: packet too short (%d) for header+IV+1block+ICV", len(pkt))
	}
	spi := binary.BigEndian.Uint32(pkt[0:4])
	seq := binary.BigEndian.Uint32(pkt[4:8])
	if spi != s.SPI {
		return nil, 0, fmt.Errorf("esp: SPI %08x != SA SPI %08x", spi, s.SPI)
	}

	// Anti-replay first per §3.4.3 — cheap check before HMAC.
	if !s.SeqInWin.check(seq) {
		return nil, 0, fmt.Errorf("esp: replay/old sequence number %d (RFC 4303 §3.4.3)", seq)
	}

	icvOff := len(pkt) - ICVLen
	icv := pkt[icvOff:]
	macInput := pkt[:icvOff]

	// HMAC-SHA-256 verify (constant-time).
	mac := hmac.New(sha256.New, s.IntegKey)
	mac.Write(macInput)
	want := mac.Sum(nil)[:ICVLen]
	if subtle.ConstantTimeCompare(want, icv) != 1 {
		return nil, 0, errors.New("esp: ICV mismatch (RFC 4303 §3.4.4 — packet rejected)")
	}

	// AES-CBC decrypt the body (after IV, before ICV).
	iv := pkt[HeaderLen : HeaderLen+IVLen]
	ct := pkt[HeaderLen+IVLen : icvOff]
	if len(ct)%aes.BlockSize != 0 {
		return nil, 0, fmt.Errorf("esp: ciphertext len %d not block-aligned", len(ct))
	}
	block, err := aes.NewCipher(s.EncrKey)
	if err != nil {
		return nil, 0, err
	}
	pt := make([]byte, len(ct))
	cipher.NewCBCDecrypter(block, iv).CryptBlocks(pt, ct)

	// Strip pad / pull NextHdr.
	if len(pt) < 2 {
		return nil, 0, errors.New("esp: plaintext shorter than PadLen+NextHdr trailer")
	}
	padLen := int(pt[len(pt)-2])
	nextHdr := pt[len(pt)-1]
	if 2+padLen > len(pt) {
		return nil, 0, fmt.Errorf("esp: pad len %d overruns plaintext %d", padLen, len(pt))
	}
	// Verify pad pattern per §2.4 — implementations SHOULD check it
	// to detect cipher errors that didn't perturb the ICV (which
	// would be the cipher itself being broken, but cheap defense).
	for i := 0; i < padLen; i++ {
		if pt[len(pt)-2-padLen+i] != byte(i+1) {
			return nil, 0, fmt.Errorf("esp: pad pattern violation at byte %d (RFC 4303 §2.4)", i)
		}
	}
	inner := append([]byte(nil), pt[:len(pt)-2-padLen]...)

	// ICV passed + replay check passed — commit the seq num to the
	// window so a duplicate packet is rejected next time.
	s.SeqInWin.commit(seq)
	return inner, nextHdr, nil
}

func appendU32(b []byte, v uint32) []byte {
	var t [4]byte
	binary.BigEndian.PutUint32(t[:], v)
	return append(b, t[:]...)
}

// replayWindow implements the §3.4.3 anti-replay window. Stores the
// last 'size' bits in a uint64 (so size <= 64). Window slides forward
// when a fresh higher seq num arrives.
//
// Verbatim from RFC 4303 §3.4.3: "If the received packet's sequence
// number falls within the window and is new, or if the packet is to
// the right of the window, then the receiver proceeds to ICV
// verification. If the ICV validation fails, the receiver MUST
// discard the received IP datagram as invalid; this is an auditable
// event."
type replayWindow struct {
	last uint32 // highest seen seq num
	size uint   // window size (typically 64)
	bits uint64 // bit i = (last - i) was seen
}

func newReplayWindow(size uint) *replayWindow {
	if size == 0 || size > 64 {
		size = 64
	}
	return &replayWindow{size: size}
}

// check returns false if seq is too old or already seen. Does NOT
// commit — caller must call commit() after the ICV passes (otherwise
// a forged packet with valid seq num but wrong ICV would shift the
// window and let a real later packet be denied).
func (w *replayWindow) check(seq uint32) bool {
	if seq == 0 {
		// §3.4.3: "the value 0 is reserved" if extended seq numbers
		// are not in use — packets with seq=0 are rejected.
		return false
	}
	if seq > w.last {
		return true // future packet, definitely accept (subject to ICV)
	}
	diff := w.last - seq
	if diff >= uint32(w.size) {
		return false // beyond window — too old
	}
	if w.bits&(1<<diff) != 0 {
		return false // already seen
	}
	return true
}

func (w *replayWindow) commit(seq uint32) {
	if seq > w.last {
		shift := seq - w.last
		if shift >= uint32(w.size) {
			w.bits = 0
		} else {
			w.bits <<= shift
		}
		w.bits |= 1
		w.last = seq
		return
	}
	diff := w.last - seq
	w.bits |= 1 << diff
}

File: tester/src/protocol/esp.py
# Copyright (c) 2026 MakeMyTechnology. All rights reserved.
"""RFC 4303 Encapsulating Security Payload — UE-side mirror.

The N3IWF UE simulator drives the data plane over NWu (TS 24.502 §7.4):
ESP-in-UDP encapsulating an inner IPv4 (NextHdr=4) or IPv6 (NextHdr=41)
packet. Cipher suite — AES-CBC-256 + HMAC-SHA-256-128 — is what the
IKEv2 handler negotiates in CREATE_CHILD_SA.

Wire layout (RFC 4303 §2 Figure 1, verbatim):
    SPI (4)
    Sequence Number (4)
    IV (16)               ⟵ AES-CBC IV
    Ciphertext (n*16)     ⟵ AES-CBC over (inner | pad | padlen | nexthdr)
    ICV (16)              ⟵ HMAC-SHA-256-128 over SPI..Ciphertext

Mirrors nf/n3iwf/esp/esp.go byte-exactly so a packet built by either
side parses on the other.
"""

import hmac as _hmac
import hashlib as _hashlib
import os
import struct
from dataclasses import dataclass, field
from typing import Tuple

from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes


HEADER_LEN = 8        # SPI(4) + SeqNum(4)
IV_LEN = 16
ICV_LEN = 16          # HMAC-SHA-256-128 (RFC 4868 §2.1)
BLOCK_SIZE = 16       # AES

# IANA "Assigned Internet Protocol Numbers" — tunnel-mode inner.
NEXT_HDR_IPV4 = 4
NEXT_HDR_IPV6 = 41


# ─── §3.4.3 Anti-replay window ────────────────────────────────────


class _ReplayWindow:
    """64-bit sliding window per RFC 4303 §3.4.3.

    'A minimum window size of 32 packets MUST be supported, but a
    window size of 64 is preferred and SHOULD be employed as the
    default.' Stores the last `size` bits in a uint64.

    check() does NOT commit — caller commits *after* the ICV passes,
    so a forged packet with a valid seq num but wrong ICV cannot
    advance the window and starve a legitimate later packet."""

    def __init__(self, size: int = 64):
        if size == 0 or size > 64:
            size = 64
        self.size = size
        self.last = 0    # highest seen seq num
        self.bits = 0    # bit i ⇒ (last - i) was seen

    def check(self, seq: int) -> bool:
        # §3.4.3: seq=0 reserved when extended seq nums are off.
        if seq == 0:
            return False
        if seq > self.last:
            return True
        diff = self.last - seq
        if diff >= self.size:
            return False
        return (self.bits & (1 << diff)) == 0

    def commit(self, seq: int) -> None:
        if seq > self.last:
            shift = seq - self.last
            if shift >= self.size:
                self.bits = 0
            else:
                self.bits = (self.bits << shift) & ((1 << self.size) - 1)
            self.bits |= 1
            self.last = seq
            return
        diff = self.last - seq
        self.bits |= (1 << diff)


# ─── SA: per-direction key + state ────────────────────────────────


@dataclass
class SA:
    """One SA per direction. Mirrors nf/n3iwf/esp/esp.go::SA.

    SeqOut starts at 0 — RFC 4303 §3.3.3: 'The sender's counter is
    initialized to 0 ... the sender increments and inserts the
    low-order 32 bits.' First transmitted seq num is 1."""

    spi: int
    encr_key: bytes              # AES-256 key (32 octets)
    integ_key: bytes             # HMAC-SHA-256 key (32 octets)
    seq_out: int = 0
    seq_in: _ReplayWindow = field(default_factory=lambda: _ReplayWindow(64))

    def __post_init__(self):
        if len(self.encr_key) != 32:
            raise ValueError(
                f"esp: encr key length {len(self.encr_key)} != 32 (AES-256)"
            )
        if len(self.integ_key) != 32:
            raise ValueError(
                f"esp: integ key length {len(self.integ_key)} != 32 "
                f"(HMAC-SHA-256, RFC 4868 §2.1)"
            )

    # ── Encap (RFC 4303 §2 + §3.3) ────────────────────────────────

    def encap(self, inner: bytes, next_hdr: int = NEXT_HDR_IPV4) -> bytes:
        """Wrap an inner IP packet. Returns the ESP packet ready to
        drop into a UDP datagram (NAT-T, RFC 3948) or raw IP-50."""
        if self.seq_out == 0xFFFFFFFF:
            raise ValueError(
                "esp: outbound sequence number wrapped (RFC 4303 §3.3.3 — must rekey)"
            )
        # Fresh, unpredictable IV. os.urandom is the cryptographic PRNG.
        iv = os.urandom(IV_LEN)
        self.seq_out += 1
        return _build_packet(
            spi=self.spi, seq=self.seq_out,
            encr_key=self.encr_key, integ_key=self.integ_key,
            iv=iv, inner=inner, next_hdr=next_hdr,
        )

    # ── Encap with caller-supplied IV / seq — TESTING ONLY ────────
    #
    # Lets the byte-vector cross-language test pin every input that
    # affects the wire bytes, so the Go and Python sides can lock to
    # the same hex constant. Production code MUST use encap(), which
    # generates a fresh random IV per RFC 4303 §3.3 ('the IV [...]
    # MUST be chosen ... to ensure that no two packets are encrypted
    # with the same IV').
    def _encap_deterministic(
        self, inner: bytes, next_hdr: int, iv: bytes, seq: int,
    ) -> bytes:
        if len(iv) != IV_LEN:
            raise ValueError(f"esp: test IV length {len(iv)} != {IV_LEN}")
        return _build_packet(
            spi=self.spi, seq=seq,
            encr_key=self.encr_key, integ_key=self.integ_key,
            iv=iv, inner=inner, next_hdr=next_hdr,
        )

    # ── Decap (RFC 4303 §3.4) ─────────────────────────────────────

    def decap(self, pkt: bytes) -> Tuple[bytes, int]:
        """Unwrap an ESP packet. Returns (inner, next_hdr).
        Raises on ICV failure, replay, or pad violation."""
        if len(pkt) < HEADER_LEN + IV_LEN + BLOCK_SIZE + ICV_LEN:
            raise ValueError(
                f"esp: packet too short ({len(pkt)}) for header+IV+1block+ICV"
            )
        spi, seq = struct.unpack(">II", pkt[:8])
        if spi != self.spi:
            raise ValueError(f"esp: SPI {spi:08x} != SA SPI {self.spi:08x}")

        # §3.4.3: replay check first — cheap before HMAC.
        if not self.seq_in.check(seq):
            raise ValueError(
                f"esp: replay/old sequence number {seq} (RFC 4303 §3.4.3)"
            )

        icv_off = len(pkt) - ICV_LEN
        icv = pkt[icv_off:]
        mac_input = pkt[:icv_off]

        want = _hmac.new(self.integ_key, mac_input, _hashlib.sha256).digest()[:ICV_LEN]
        # §3.4.4 — constant-time compare.
        if not _hmac.compare_digest(want, icv):
            raise ValueError("esp: ICV mismatch (RFC 4303 §3.4.4 — packet rejected)")

        iv = pkt[HEADER_LEN:HEADER_LEN + IV_LEN]
        ct = pkt[HEADER_LEN + IV_LEN:icv_off]
        if len(ct) % BLOCK_SIZE != 0:
            raise ValueError(f"esp: ciphertext len {len(ct)} not block-aligned")

        pt = _aes_cbc_decrypt(self.encr_key, iv, ct)
        if len(pt) < 2:
            raise ValueError("esp: plaintext shorter than PadLen+NextHdr trailer")
        pad_len = pt[-2]
        next_hdr = pt[-1]
        if 2 + pad_len > len(pt):
            raise ValueError(
                f"esp: pad len {pad_len} overruns plaintext {len(pt)}"
            )
        # §2.4 pad pattern check — defends against cipher-error packets
        # that somehow survive the ICV.
        for i in range(pad_len):
            expected = i + 1
            actual = pt[-2 - pad_len + i]
            if actual != expected:
                raise ValueError(
                    f"esp: pad pattern violation at byte {i} (RFC 4303 §2.4)"
                )
        inner = bytes(pt[:-2 - pad_len])

        # ICV + replay both passed — only NOW commit so a forged packet
        # can't shift the window.
        self.seq_in.commit(seq)
        return inner, next_hdr


# ─── helpers ──────────────────────────────────────────────────────


def _build_packet(
    spi: int, seq: int,
    encr_key: bytes, integ_key: bytes,
    iv: bytes, inner: bytes, next_hdr: int,
) -> bytes:
    """The pure transformation: (inputs) → wire bytes.

    No state mutation, no IV generation, no seq-num bump — caller
    owns both. Shared by encap() (production path with random IV +
    auto-incremented seq) and _encap_deterministic() (test path
    that pins both for byte-exact cross-language vectors)."""
    if not inner:
        raise ValueError("esp: inner payload empty")
    # Plaintext = inner | padding | padlen | nexthdr.
    # Total must be a multiple of AES block size. Pad pattern per
    # §2.4: 1, 2, 3, ..., padlen.
    tail_len = 2  # PadLen + NextHdr
    rem = (len(inner) + tail_len) % BLOCK_SIZE
    pad_len = 0 if rem == 0 else (BLOCK_SIZE - rem)
    if pad_len > 255:
        raise ValueError(f"esp: pad length {pad_len} > 255 (RFC 4303 §2.4)")
    plaintext = bytearray(inner)
    for i in range(pad_len):
        plaintext.append(i + 1)             # §2.4 pattern
    plaintext.append(pad_len)
    plaintext.append(next_hdr)

    ct = _aes_cbc_encrypt(encr_key, iv, bytes(plaintext))

    pkt = struct.pack(">II", spi, seq) + iv + ct
    # §2.8: ICV scope is SPI..Ciphertext (not including ICV itself).
    icv = _hmac.new(integ_key, pkt, _hashlib.sha256).digest()[:ICV_LEN]
    return pkt + icv


def _aes_cbc_encrypt(key: bytes, iv: bytes, plaintext: bytes) -> bytes:
    c = Cipher(algorithms.AES(key), modes.CBC(iv)).encryptor()
    return c.update(plaintext) + c.finalize()


def _aes_cbc_decrypt(key: bytes, iv: bytes, ciphertext: bytes) -> bytes:
    c = Cipher(algorithms.AES(key), modes.CBC(iv)).decryptor()
    return c.update(ciphertext) + c.finalize()


def peek_spi(pkt: bytes) -> int:
    """Demux helper — first 4 octets are SPI per RFC 4303 §2."""
    if len(pkt) < 4:
        raise ValueError("esp: packet too short to contain SPI")
    return struct.unpack(">I", pkt[:4])[0]
